[depends] Qt 5.7.1

fanquake opened the pull request:
This updates the depends system to use Qt [5.7.1](https://blog.qt.io/blog/2016/06/16/qt-5-7-released/). One improvement includes:
```
Starting with Qt 5.7, we will require a C++11 compliant compiler to build and use Qt. 
This allows us to use many of the new features in Qt itself, and renew our codebase 
using features such as constexpr, move semantics, range-for and auto. These efforts 
are well under way and will continue throughout the next versions.
```

See #8237 for some more discussion. However it's worth noting that build-related improvements likely won't make it into 0.14.0.

We can drop the configure-xcoderun patch, as it's been fixed in src.
The other patches have been updated to apply cleanly.

-c++11 and -no-nis have been removed as they are now invalid options.

There is currently an lrelease/translations related issue, and one other on linux.
```
compiling ../shared/qmakeevaluator.cpp
compiling ../shared/qmakebuiltins.cpp
compiling ../shared/profileevaluator.cpp
linking ../../../bin/lrelease
make[1]: Leaving directory `/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-unknown-linux-gnu/qt/5.7.1-b0275a68b64/qttools/src/linguist/lrelease'
make[1]: Entering directory `/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-unknown-linux-gnu/qt/5.7.1-b0275a68b64/qttranslations'
cd translations/ && ( test -e Makefile || /home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-unknown-linux-gnu/qt/5.7.1-b0275a68b64/qtbase/bin/qmake /home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-unknown-linux-gnu/qt/5.7.1-b0275a68b64/qttranslations/translations/translations.pro -o Makefile ) && make -f Makefile 
make[2]: Entering directory `/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-unknown-linux-gnu/qt/5.7.1-b0275a68b64/qttranslations/translations'
make[2]: *** No rule to make target `/home/travis/build/bitcoin/bitcoin/depends/x86_64-unknown-linux-gnu/native/bin/lrelease', needed by `assistant_cs.qm'.  Stop.
```

cc @theuni 
Closes #8237

fanquake committed:
[depends] Qt 5.7.1

paveljanik commented on depends/patches/qt/fix_qt_pkgconfig.patch:
```diff
@@ -1,6 +1,6 @@
---- old/qtbase/mkspecs/features/qt_module.prf 2016-03-17 02:06:42.705930685 +0000
-+++ new/qtbase/mkspecs/features/qt_module.prf 2016-03-17 02:06:42.705930685 +0000
-@@ -244,7 +244,7 @@
+--- old/qtbase/mkspecs/features/qt_module.prf
```
Why so many diff-only realted changes here?

laanwj replied:
The code stayed the same but the lines moved.

paveljanik replied:
We already clarified this with @fanquake. He re-did the diffs to apply cleanly.

theuni commented:
@fanquake Thanks, having a look.

theuni commented:
@fanquake Seems the xcoderun issue still exists. Did you see something upstream that looked like a fix?

I remember fixing up some of those issues in my unfinished branch, which was an attempt to split up the monolithic build into smaller chunks: https://github.com/theuni/bitcoin/commit/b00b202690322917df6927997b569587af31c438. It's obviously not worth trying to do all that now, but it may be helpful to pick some of the changes out.

IIRC the lrelease problem was caused by something assuming that lrelease was already built, forcing it to build earlier should do the trick.

jtimon commented:
Concept ACK

fanquake commented:
@theuni I thought it only affected the 5.6 branch https://codereview.qt-project.org/#/c/164673.
Homebrew also dropped their patch for the same issue when moving to 5.7.1, see this [commit](https://github.com/Homebrew/homebrew-core/commit/1c4bb08cc9cff85d0ae921886263064ef8bd3b30).

theuni commented:
@fanquake That's a different issue. Our problem is that it assumes that builds for osx are done on osx. Our patch fixes the cross build.

theuni committed:
depends: use new variable layout for qt sdk

theuni committed:
depends: fix qt translations build

Their buildsystem insists on using the installed ltranslate, but gets confused
about how to find it. Since we manually control the build order, just drop the
dependency.

theuni commented:
@fanquake My apologies, you were right about that patch. I had something different in mind (a sed we do in the .mk).

Mind putting that back the way you had it, then grabbing the top two commits here: https://github.com/theuni/bitcoin/tree/qt-5.7new ?

I think that may be enough to get this working.

theuni commented:
@fanquake Also, rather than dropping the "-c++11" option, let's change it to "-c++std c++11". Otherwise, it'll pick up c++14/c++17 and potentially end up with a confused abi.

fanquake commented:
Rebased, so this is now on top of the other depends changes.
Updated to use `-c++std c++11` instead of just dropping `-c++11`.
Re-dropped the xcrun configure patch.
Pull in two commits from @theuni.

fanquake commented:
After discussion with theuni on IRC, have just pushed a fix for the c++ issue with the OSX build. 

If this works, make sure to split out the config.site.in change into a separate commit, as it could be backported.

```
cfields: fanquake only reason i mention is that i had a theory at one point that it was the cause of the 10.7 back-compat breakage. Since it's responsible for intoducing c++03 abi objects into the 0.13 build.
```
https://github.com/bitcoin/bitcoin/blob/master/configure.ac#L62

fanquake committed:
[depends] Remove OBJCXX define from config.site.in

MarcoFalke commented:
Concept ACK.

laanwj commented:
utACK bb077fa, looks good to me, good to see this is passing now.

MarcoFalke commented:
> If this works, make sure to split out the config.site.in change into a separate commit, as it could be backported.

Tagging for backport
